add payment channel contracts with virtual channel support

ravi0131 opened the pull request:
This PR introduces the contracts for perun-channel and virtual channel. 

Features:
+ contracts for payment channel (PCTS, PCLS, PFLS)
+ contracts for on-chain dispute resolution of virtual channels (PCTS, PCLS)
+ tests

Commit:
tests: Implement a test harness encapsulating chain env

Commit:
tests: Add abstract channel actions

Commit:
tests: Add a ChannelState mock for use in tests

Commit:
tests: Prepare a test module containing various stub impls

Commit:
tests: Make use of new perun module

Commit:
tests: Setup testbench and adapt to Perun contracts

Commit:
tests: Add rand dependency to tests

Commit:
tests: Give channel_id a random constructor

Commit:
tests/harness: Allow Env to build multiple perun scripts

Commit:
tests/channel: Add a constructor for Channel

Commit:
tests: Derive Debug, Clone for Client and boyscout compiler err

Commit:
common: Make std and no_std compatible

Commit:
contracts: Explicitly use no_std version for perun-common

Commit:
lock: Update lockfile

Commit:
tests/Cargo: Update ckb-testtool dependency to 0.9

Commit:
common: Implement macro helpers for constructing boolean values

Commit:
tests: Implement client open functionality

janbormet committed:
common: Add helper to convert from boolean to on-chain bool

janbormet committed:
tests: Add test for signature verification

janbormet committed:
common: Remove comment that signature verification is just a draft

janbormet committed:
pcls: Add documentation

janbormet committed:
pcts: Add some documentation for ChannelAction and clean up code

janbormet committed:
pcts+common: Verify that there is at most one channel present in the transaction

janbormet committed:
common: Add the minimum capacity of the pfls and the payout script to the channel constants

janbormet committed:
pcts: Allow not to pay out parties whose balance is below the minimum capacity of the payout script

janbormet committed:
common: Use macros for from_bool

janbormet committed:
pcts+common: Verify upon start that no individual balance is below the minimum capacity of the pfls

janbormet committed:
pcts+common: Allow each participant to specify different payment scripts and verify that the payment addresses defined by the payment script and payment args of each participant are different

Commit:
common: Boyscout formatting

Commit:
tests/tests: Resolve conflicting imports

janbormet committed:
common: Fix bug in boolean to Bool conversion

janbormet committed:
pcts: Allow a party to start the channel with 0 balance

janbormet committed:
pcts: Document channel start

janbormet committed:
pcts+common: Verify that the sum of the balances are equal upon dispute instead of the actual balances

Commit:
tests/Cargo: Add ckb-occupied-capacity dependency

janbormet committed:
pcts: Document channel progression

Commit:
tests: Signify to compiler that test crates are only used in tests

Commit:
tests/account: Add an account abstraction

Commit:
tests/random: Implement random helpers

Commit:
tests/perun: Move TX building into own module

Commit:
tests/error: Allow conversion between Capacity::Error and perun::Error

Commit:
tests/perun: Finish client impl for open case

Commit:
tests: Adapt DSL usage to new API

janbormet committed:
pcts: Document channel close

janbormet committed:
pcts: Fix bug in counting channel outputs

Commit:
tests: Calculate ChannelID when opening a channel in testsuite

Commit:
tests: Hash the ChannelParams not ChannelConstants for id calc

janbormet committed:
pcts: Add debug statements and fix get_own_index

janbormet committed:
pcts: Fix misuse of load_cell_type_hash

janbormet committed:
pcts: Use GroupInput as source instead of manually locating own index in inputs

janbormet committed:
pfls: Fix misuse of load_cell_type_hash

janbormet committed:
pcts: Use GroupInput and GroupOutput instead of manually locating the index of the pcts

janbormet committed:
tests: Increase balances in funding agreement to be larger thant the pfls minimum capacity (hardcoded)

janbormet committed:
pcts: Clarify comment

Commit:
tests/error: Implement From<ckb_error::Error> for perun::Error

janbormet committed:
contracts: Distinguish strictly between "script hash" and "code hash"

We want to use a script hash, whenever we want to compare scripts, as the script hash includes
the code hash, the hash type (which we just ignored before this) and the args.

The following members are now script hashes:
- payment_script_hash of a participant (specifies which script to use to pay out a participant
- the args of the pfls (this is now the script hash of the type script of the channel to which
	the funds are locked
- unlock_script_hash of a participant

The following member are now code hashes (together with their hash type):
- pfls_code_hash: specifies which pfls to use
- pcls_code_hash: specifies which pcts to use
Here, we do not want to use script hashes, as this would introduce cyclic dependencies of hash
inputs.

The funding and payout checks, as well as the integrity checks on the channel constants were
adjusted accordingly

Commit:
Merge remote-tracking branch 'origin/debug' into extend-testsuite

Commit:
tests: Cache and thread channel_outpoint and funds after open action

Commit:
common: Add a macro for channel action/redeemer construction

Commit:
tests: Use explicit code/type_hash qualified member names

Commit:
tests: Adapt to new perun_types API

Commit:
tests: Add FundingCell abstraction w/ meta infos about funding party

Commit:
tests: Impl abort functionality in test-suite

Commit:
tests: Adapt channel opening construction to updated hash usage

Commit:
common/helpers: Impl/Extend macros for fund and witness construction

Commit:
common/helpers: Impl From<u8> for ParticipantIndex

Commit:
common/helpers: Give ChannelStatus aux func to update funding/funded

Commit:
tests: Impl fund functionality in test-suite

Commit:
common/helpers: Fully qualified paths in macros

Commit:
tests/transaction: Modularize transaction implementations

Commit:
common/helpers: Add aux methods for ChannelStatus/Balances

Commit:
tests/error: Impl From<k256::ecdsa::Error> for perun::Error

Commit:
tests/transaction/close: Pass cell_deps to built TX

Commit:
tests/client: Implement sign method for clients

janbormet committed:
pcts+common: Change Balance array to underlying type Uint64 from Uint128

janbormet committed:
tests: Change Balance array to underlying type Uint64 from Uint128

Commit:
tests: Add molecule as a dependency for tests

Commit:
common/helpers: Add convenience macro for Close redeemer

Commit:
common/helpers: Pass mutable closures for Output Script creation

Commit:
common/helpers: Aux methods for ChannelParams to fetch party pubkeys

Commit:
tests/channel: Implement close functionality

Commit:
tests/client: Implement close functionality

Commit:
tests/transaction: Implement close transaction constructor

Commit:
tests/error: Add convenience From<_> impls for perun::Error

Commit:
Merge remote-tracking branch 'origin/debug' into extend-testsuite

Commit:
tests: Add traces for close, force_close and proper channel conclusion

Commit:
tests/transaction: Add ForceClose transaction implementation

Commit:
tests/channel: Add force_close implementation to test-suite

Commit:
tests/client: Give client ability to create ForceClose transaction

Commit:
tests: Boyscout imports

Commit:
tests/client: Use correct payment_script_hash for parties

Commit:
tests/transaction: Add dispute transaction implementation

Commit:
common/helpers: Add convenience macro for Dispute redeemer

Commit:
tests/client: Add ability to create Dispute transaction

Commit:
tests/channel: Add dispute implementation to test-suite

janbormet committed:
pcts+pcls+common: Apply review suggestions from Ian

This applies the following suggestions:
- Fix bug in `verify_different_payment_addresses`
- Now we verify that there is at most one channel in the GroupInput/GroupOutput respectively
	by efficiently counting the cells in the respective sources because CKB already groups
	by type script and script args. This potentially allows to interact with in the same
	transaction (even to be started in the same trasaction), but this should not be a problem
	as long as their script-hashes differ (which we still verify)!
- Optimize `get_channel_action`
- Clarify doc in pcls
- Improve variable naming in `verify_all_payed` and `verify_funding_unchanged`
- Do not check the presence of a matching pcts again in `verify_channel_continues`. As this was
	already asserted in previous checks, we only verify that the continuing channel cell is
	locked by the same lock script. This also removes the `is_matching_output` helper.
- Verify that the args field of the pcls is empty upon channel start

Commit:
Merge remote-tracking branch 'origin/debug' into extend-testsuite

Commit:
tests/transaction: Pass header_deps to all relevant txs

Commit:
tests/channel: Make sure to insert_header into ctx before linking cells

Commit:
common/helpers: Let mk_close_outputs also return outputs data for cells

Commit:
tests/transaction: Pass outputs_data to abort, close, force_close txs

Commit:
tests/transaction: Create output cells for channel txs in Context

Commit:
tests: Pass a payment input for acting party to each transaction

Commit:
tests/transaction: Pass minimum capacity for channel to close txs

Commit:
tests/client: Give client a name identifier for tests

Commit:
tests/channel: Print the action that is executed in tests

Commit:
pcls: Verify participants for both parties

Commit:
tests/harness: Boyscout DRY style

Commit:
tests/keys: Correctly decode VerifyingKey to bytearray

Commit:
tests/channel: Properly recode raw molecule bytes off-chain

Commit:
tests/channel: Use perun::TestAccount as a source for SigningKeys

Commit:
tests/channel: Add and use finalize helper for ChannelState

janbormet committed:
common: Use 33 bytes compressed SEC1 encoding for public keys

janbormet committed:
tests: Use 33 bytes compressed SEC1 encoding for public keys

janbormet committed:
pcts: Fix error in comment

Commit:
tests/channel: Valid disputes using a new state

Commit:
tests/tests: Bump MAX_CYCLE_NUMBER in testsuite

janbormet committed:
pcts+common: Move hashing the message out of the signature verification function

While this clearly belongs inside the signature verification function, it is better
to move this into the caller, because our pcts always verifies two signatures on the
same message. Thereby we avoid calculating the same hash twice.

Commit:
tests/tests: Simplify multiple dispute tests for happy case

janbormet committed:
tests: Calculate the message hash before verifying the signature

janbormet committed:
pcts: Add consistent debugging output

janbormet committed:
capsule: Bump capsule version to 0.9.2

janbormet committed:
Merge pull request #7 from perun-network/extend-testsuite

Extend Test-Suite

janbormet committed:
workspace: Exclude tests from cargo workspace

janbormet committed:
workspace: Apply 'perun-common-should-default-for-contract'

janbormet committed:
deployment: Add deployment configuration

janbormet committed:
Merge pull request #8 from perun-network/workspace-setup

workspace: Exclude tests from cargo workspace

janbormet committed:
common: Remove the ParticipantIndex type

The funding protocol we use does not strictly need to specify an
index in the Fund witness, as only the 2nd party will ever call fund.
Therefore we remove this type to avoid unnecessary complexity.

janbormet committed:
pcts: Remove ParticipantIndex

janbormet committed:
tests: Remove ParticipantIndex

janbormet committed:
Merge pull request #9 from perun-network/remove-participant-index

Remove participant index

janbormet committed:
pcts+common: Add SUDT Asset support

janbormet committed:
tests: Adjust reference pubkey and signature to new Balance type in signature test

janbormet committed:
tests/funding_agreement: Hack funding agreement to CKBytes funding (temporary)

janbormet committed:
common: Add OffChainParticipant type for go backend

janbormet committed:
pcts+common: Remove unnecessary Funding allocation from ChannelStatus type

janbormet committed:
tests: Prepare test framework for sudt tests

janbormet committed:
pcts: Fix bug regarding increasing version number and remove unused code

janbormet committed:
tests: Add ChannelState mutation to DSL

janbormet committed:
pcts+common: Clean up code

janbormet committed:
sample-udt: Add sample udt for testing

janbormet committed:
pcts: Do not assume that the channel cell is the first cell in the inputs / outputs

janbormet committed:
pcts: Add some debug prints to verify_all_payed

janbormet committed:
pcts: Unpack Uint64 in debug print

janbormet committed:
common: Allow for generic outputs generation from Balances

janbormet committed:
tests: Rename mutator nop to id

janbormet committed:
tests: Add functionality to FundingAgreement and AssetRegister

janbormet committed:
tests: Allow for CKBytes and SUDT funding cells

janbormet committed:
tests: Use pcls code hash instead of recomputing

janbormet committed:
tests: Add utility for adding some capacity to party a's balance

janbormet committed:
tests: Rework input and output generation for open, fund, abort, close and force-close transaction building

janbormet committed:
tests: Add sample udt script to env

janbormet committed:
tests: Adjust tests for UDTs and add more tests

janbormet committed:
common: Add to_array / from_array helpers to CKByteDistribution and SUDTDistribution

janbormet committed:
tests: Add doc clarifying that finalize bumps the version number

janbormet committed:
tests: Add multi-asset payment test

janbormet committed:
pcts: Fix bug in payout verification

janbormet committed:
tests: Add multi-asset abort tests

janbormet committed:
common: Fix bug in construction of sudt outputs for tests and format code

Before this, we failed to include an output for an sudt asset, if the amount
in the asset's distribution was 0. We need to include it though, because of the
sudt's max capacity (it needs to be funded and refunded).

janbormet committed:
pcts: Fix bug regarding sudt payout

janbormet committed:
tests: Add multi-asset force close test

janbormet committed:
common: Fix bug in balance sum equality

janbormet committed:
pcts: Format code

Commit:
Create rust.yml

Commit:
.gitignore: Add tests/target folder

Commit:
CI: Setup CI with capsule for contracts and perun-common library

Commit:
README: Add CI and license batches

janbormet committed:
Merge pull request #11 from perun-network/capsule-ci

Capsule ci

Commit:
Merge pull request #10 from perun-network/sudt-support

Sudt support

Commit:
Update README.md

doitian committed:
pcls: Loop to verify all channel locks

Commit:
deployment: Also deploy SUDT for debugging purposes

Commit:
deployment: Create separate deployment files for release and development

Commit:
capsule.toml: Reference dev/deployment per default

Commit:
Merge pull request #12 from perun-network/pcls-script-groups

Verify all channel locks in script group in pcls

iljabvh committed:
Refactor (WIP): Represent VCs with its own Lockscript and Typescript

iljabvh committed:
Adapt ChannelAction for VCs to having a unique Typescript and Lockscript.

* chore(contracts): Fix typo

* refactor(contracts): Extend ChannelAction struct to reflect changes on VC cell creation and termination. Adapt main() function.

* chore(contracts): Add custom errors

iljabvh committed:
Refactor & Test WIP: Adapt Code to new VC architecture

iljabvh committed:
Refactor(contracts):
Move helper functions from entry.rs to channels.rs

NhoxxKienn committed:
feat: update cargo

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
feat: add vc scripts to env

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
feat: vc setup for test (WIP)

Signed-off-by: Minh Huy Tran <huy@perun.network>

iljabvh committed:
Fix(perun-vchannel-typescript):
Correct case for Dispute start and add build_vchannel_status function to initialize a VC state for a VC cell, given data from PCTS.

NhoxxKienn committed:
WIP: fix compilation bugs and add simple open_vc mechanism

Signed-off-by: Minh Huy Tran <huy@perun.network>

iljabvh committed:
* fix(types.mol): Remove ChannelStatus vector from VirtualChannelStatus type
* refactor(contracts): Move utility functions from entry.rs files to channels.rs, remove comments

NhoxxKienn committed:
add closing virtual channel and fix some compilation errors (WIP)

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
feat: refactor vc happy test

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
feat: add vc_channel_happy_test

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
feat: add interior mutability for context

Signed-off-by: Minh Huy Tran <huy@perun.network>

ravi0131 committed:
WIP: Contract changes

ravi0131 committed:
feat: VC Support

Changes made to contract to support vc cells
* open vc cell
* progress without update
* progress with update
* vc merge
* vc close 1
* vc close 2

ravi0131 committed:
Resolved merge conflict in types.mol

ravi0131 committed:
fix: update types and remove comments

* update types.mol and change contract for the updated types.
* chore: remove comments

ravi0131 committed:
chore: remove unneccsary doc

ravi0131 committed:
Merge branch 'feat-vc-vcell' into vc_tests

ravi0131 committed:
chrore: resolve merge conflict after stash apply

ravi0131 committed:
WIP: fix(virtual_channel.rs): fix update and resolve functions for virtual channel

- fix(virtual_channel.rs): update_virtual_channel() to correctly lock the correct amounts
- fix(virtual_channel.rs): resolve_virtual_channel() to correctly release locked funds
- chore(cargo.toml): add tests to cargo.toml in project root to enable syntax support for tests.

ravi0131 committed:
feat: add VC Dispute Start Test

- tests.rs : main test for vc dispute start (test_vc_start)
- channel.rs, client.rs: add supporting methods
- vc_start.rs: add types to hold information for tx and methods to create tx
- virtual_channel.rs : add type to represent off_chain state of vc and corresponding methods
- harness.rs, helpers.rs: add functionality to support vc start test

ravi0131 committed:
refactor: test_vc_start  && fix:  channel_test_bench

refactor(tests.rs, virtual_channel.rs, channel.rs): refactor test_vc_start for more cleaner code

fix(channel.rs, harness.rs): remove deadlock and correctly create initial channel cell (set vc_dispute flag to false)

refactor (harness.rs): getter methods for vcts and vcls

ravi0131 committed:
feat: add test_vc_progress_no_update

ravi0131 committed:
feat: add vc_update_only && fix: dispute type molecule

- add test for only updating vc cell
- fix (types.mol): add sig_b field in dispute

sophia1ch committed:
feat: generated test file for ckb-scripts and Makefiles
chore: renamed files and updated dependencies
note: building the contract works but running the tests does not yet

ravi0131 committed:
feat: add vc merge test

sophia1ch committed:
refactor: regenerated ckb workspace and updated all dependencies
feat(NOTED): added issues when building and testing contracts

sophia1ch committed:
refactor: move perun-common to crates directory
fix(perun-channel-typescript/lib.rs): rewrite unpack to handle different implemetations of bytes32 and u64

ravi0131 committed:
fix: update outpoints after tx

update the outpoint after sending tx to dummy chain and set headers for vc cell

vc_update_only.rs: remove unnecessary types

sophia1ch committed:
feat: added channel_test back in
fix: properly link lib.rs files in main
note: mak build and make test work but tests fail with ValidationError

sophia1ch committed:
feat(harness): debugging statements and use build_script_with_hash_type instead of build_script.
fix(tests): remove generated tests
fix(sample_udt): previously copied wrong code

sophia1ch committed:
fix(lib.rs): use .unpack() again

sophia1ch committed:
fix: tests can now run from terminal using makefile

sophia1ch committed:
fix: add files

ravi0131 committed:
fix: test_vc_progress_update1

entry.rs: fix verify_vc_sigs_progress and use correct source in get_vchannel_action()

vc_update_no_progress.rs: remove print statements

ravi0131 committed:
feat: add vc close 1 test

ravi0131 committed:
fix (vc_close1): use idx maps with direction

use type IdxMapWithDirection in vc_close1 instead of a simple array.

ravi0131 committed:
feat: add vc_close2

vc_close1.rs : correct output vc cell's outpoint
channel.rs: vc_close2() and formatting
client.rs:   vc_close2()
mod.rs, vcl_close2.rs, tests.rs. rs: add types and functions to support vc_close2 test

ravi0131 committed:
feat: add sudt and other tests

- test_vc_progress_update2: test updates for both lc and vc in a single tx

- test_vc_happy: test life-cycle of a vc cell without merge

- test_vc_happy_multi_asset: test life cycle of vc cell for ckb  and sudt (without merge)

- test_vc_happy_with_merge: test life-cycle of a vc cell with merge

-test_vc_happy_multi_asset_with_merge: test life-cycle of a vc cell for multiple assets (ckb +sudt) with merge tx

ravi0131 committed:
chore: cleanup

- rename vc_update_no_progress to vc_progress_no_update
- rename vc_state to vc_status in for all tx (test)
- remove vcls_script from args type where it is not needed (test)
- remove unneccessary test cases (test_vc_fund_close)
- remove debugging statements and unnecessary comments from contracts (vcts, pcts)
- add comments ( perun-common::channel.rs)

ravi0131 committed:
chore: enforce consistent code style using rustfmt

NhoxxKienn committed:
feat: add vc scripts to deployment.toml

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
fix: reorder deployment cells to fit with older deployment

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
feat(error): add new error types for vchannel

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
fix: return the sudt cell to correct position

Signed-off-by: Minh Huy Tran <huy@perun.network>

sophia1ch committed:
feat: prerequisites in readme

ravi0131 committed:
feat: add Problems section in readme

ravi0131 committed:
fix: update vc_progress logic

Add logic to prevent vc_status from being modified when it is not actively updated.

ravi0131 committed:
feat: add owner field and corresponding logic

- update vcts to check that the participant who put up funds to create the vc_cell gets reimbursed at the end
- update tests

NhoxxKienn committed:
chore: clean up contracts's unused code

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
chore: clean up tests's code

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
fix: update signature test with new channel state definition

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
chore: restore

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
feat: update README and fix variable name vcstate in test

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
chore: remove unnecessary logging

Signed-off-by: Minh Huy Tran <huy@perun.network>

ravi0131 committed:
Merge pull request #16 from perun-network/vc_tests

feat: extend contracts to support virtual channels

ravi0131 committed:
fix: failing tests after script migration

NhoxxKienn committed:
feat(deployment): update deployment order

Signed-off-by: Minh Huy Tran <huy@perun.network>

NhoxxKienn committed:
chore(CI): update cache actions dependency

Signed-off-by: Minh Huy Tran <huy@perun.network>

ravi0131 committed:
feat: add support for virtual channels

- add logic for virtual channels
- tests

ravi0131 committed:
chore: run cargo fmt && add type annotations for compiler

ravi0131 committed:
chore: remove comments

ravi0131 committed:
chore: resolve merge conflicts

ravi0131 committed:
chore: update github workflow

ravi0131 committed:
fix: deployment for vc
- vc contracts have a separate .toml file
- vc contracts have a separate migrations folder

ravi0131 committed:
chore: update readme.md

- update readme.md
- remove notes.md

ravi0131 committed:
Merge pull request #17 from perun-network/ckb-script-migration

doitian reviewed (commented):
Does the contract allow disputing/closing/force-closing a channel still having locked balances for virtual channels? I cannot find the logic in the code. The only place to check the `locked` fields in the channel balance is `check_vc_dispute`. How it impact the virtual channel when one of the parent channel has been closed? Say Alice and Ingrid has a channel (Alice: 0, Ingrid: 0, locked: (Alice: 1, Ingrid: 100)), Ingrid and Bob has a channel (Ingrid: 0, Bob: 0, locked: (Ingrid: 1, Bob: 100)), Alice and Bob has a virtual channel (Alice: 1, Bob: 100). Can Ingrid and Bob get their funds 101 back by closing their channel without closing/disputing the virtual channel? If Bob has sent 50 to Alice via the virtual channel, and the channel between Ingrid and Bob has closed, can Ingrid dispute the channel between him and Alice with the first version to get his 100 back?

doitian commented:
For reviewers: major changes are in https://github.com/perun-network/perun-ckb-contract/pull/16

ravi0131 committed:
fix: prevent closing lc channels with locked funds

* prevent closing and disputing lc channels
* allow transition from normal dispute to vc_dispute
* add check for vc flag

ravi0131 commented:
@doitian Thank you for your feedback. You’re correct. The contract previously lacked logic to prevent the closure of channels with locked funds. This created a risk in scenarios where a parent ledger channel was closed before its associated virtual channel, potentially resulting in the loss of locked funds.

I've updated the contract logic to address this issue and added test cases to ensure the behavior is correctly enforced. With this logic the following happens in regards to the scenario described by you.

`Can Ingrid and Bob get their funds 101 back by closing their channel without closing/disputing the virtual channel?`
Only if both Ingrid and Bob agree not to register the latest state that includes locked balances (which would indicate the existence of a virtual channel). Else if either of them registers an older state without locked balances then the other can register a higher state with locked balances.
(higher state = state with higher version number, same for lower state)

`Bob has sent 50 to Alice via the virtual channel, and the channel between Ingrid and Bob has closed, can Ingrid dispute the channel between him and Alice with the first version to get his 100 back?` 
No because Alice has the higher state where Bob sent her 50. So Ingrid will not get 100 back but rather only 50.

sophia1ch committed:
feat: eth signing

iljabvh committed:
Feat (WIP): Cross-chain Ethereum support

* Extension of Participant type with eth_address field in perun/common/types.mol

* Addition of alloy-* crates inside tests/Cargo.toml

* Addition of tests/src/perun/sol.rs introduces Solidity-based encoding of CKB types.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Feat(types.mol, perun_types.rs) WIP: Inclusion of ETH related assets

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Feat(src/perun/test):
Include ETHAsset into funding agreement object to account for Ethereum-based cross-chain assets

Test(tests/src/tests.rs):
Introduce test_successful_funding_with_udt_and_eth test

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Feat: ChannelId calculation adapted to Ethereum contract for cross-chain payment channels

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

ravi0131 committed:
feat: add support for type-id

ravi0131 committed:
Merge pull request #20 from perun-network/feat-type-id

feat: add support for type-id

ravi0131 committed:
fix: allow dummy entry in locked balances

ravi0131 committed:
Merge pull request #22 from perun-network/fix-locked-funds-closing

fix: allow dummy entry in locked balances

iljabvh committed:
Chore(perun_types.rs): Update to moleculec 0.8.0 and rebuild

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Refactor(perun-common): Remove EthAddress from Participant struct. The Ethereum address is retrieved from the pub_key field. Adapt tests to this change.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Fix(sol.rs): Replace manual linear byte encoding with CKB Molecule encoding

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Feat: Encode ChannelState ETH-style,  making CKB States fully compatible to the Ethereum Perun Contract and Backend.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Fix(setup_env.sh): Rename setup_env.sh to setup_env_global.sh and reintroduce the original setup_env.sh. Depending on the installation of the riscv64 crates, either script should work setting up the environment.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Fix(sol.rs): Substitute blake2b256 hashing with Keccak256 Hashing for ChannelID calculation, adapt tests accordingly.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Fix(perun-common): Substitue blake2b256 Hash function with Keccak256 , add CKBYTE_MAGIC and SUDT_MAGIC prefixes to serialization of CKB-based assets

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

sophia1ch committed:
fix: rename max_capacity to chain_id

sophia1ch committed:
Merge remote-tracking branch 'origin/ckb-eth-cross' into ckb-eth-cross

sophia1ch committed:
fix: EthAsset encoding
fix: error handling -> errors are no longer overwritten with error code -1

sophia1ch committed:
feat: keep ordered assets in balances

sophia1ch committed:
Merge pull request #19 from perun-network/ckb-eth-cross

Introduction of Ethereum-based encoding

sophia1ch committed:
fix: add ethereum description

sophia1ch committed:
fix: use union for Asset types
feat: added test coverage

sophia1ch committed:
Merge branch 'refs/heads/dev' into ckb-eth-sign

# Conflicts:
#	deployment/dev/deployment_0.toml

sophia1ch committed:
fix: enable type id

sophia1ch committed:
fix: delete generated go code

sophia1ch committed:
fix: delete generated go code

sophia1ch committed:
fix: handle eth balances

sophia1ch committed:
CKB Cross-Chain (#23)

* feat: eth signing

* Feat (WIP): Cross-chain Ethereum support

* Extension of Participant type with eth_address field in perun/common/types.mol

* Addition of alloy-* crates inside tests/Cargo.toml

* Addition of tests/src/perun/sol.rs introduces Solidity-based encoding of CKB types.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Feat(types.mol, perun_types.rs) WIP: Inclusion of ETH related assets

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Feat(src/perun/test):
Include ETHAsset into funding agreement object to account for Ethereum-based cross-chain assets

Test(tests/src/tests.rs):
Introduce test_successful_funding_with_udt_and_eth test

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Feat: ChannelId calculation adapted to Ethereum contract for cross-chain payment channels

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Chore(perun_types.rs): Update to moleculec 0.8.0 and rebuild

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Refactor(perun-common): Remove EthAddress from Participant struct. The Ethereum address is retrieved from the pub_key field. Adapt tests to this change.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Fix(sol.rs): Replace manual linear byte encoding with CKB Molecule encoding

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Feat: Encode ChannelState ETH-style,  making CKB States fully compatible to the Ethereum Perun Contract and Backend.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Fix(setup_env.sh): Rename setup_env.sh to setup_env_global.sh and reintroduce the original setup_env.sh. Depending on the installation of the riscv64 crates, either script should work setting up the environment.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Fix(sol.rs): Substitute blake2b256 hashing with Keccak256 Hashing for ChannelID calculation, adapt tests accordingly.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* Fix(perun-common): Substitue blake2b256 Hash function with Keccak256 , add CKBYTE_MAGIC and SUDT_MAGIC prefixes to serialization of CKB-based assets

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

* fix: rename max_capacity to chain_id

* fix: EthAsset encoding
fix: error handling -> errors are no longer overwritten with error code -1

* feat: keep ordered assets in balances

* fix: add ethereum description

* fix: use union for Asset types
feat: added test coverage

* fix: enable type id

* fix: delete generated go code

* fix: delete generated go code

* fix: handle eth balances

---------

Signed-off-by: Ilja von Hoessle <ilja@perun.network>
Co-authored-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
fix: replace panic with error return, optimize release profile for smaller binaries

- Replace panic!() in get_idx_map() with Err(Error::TypeHashNotFound)
- Remove stale #[allow(dead_code)] on eth_rows() which is now used
- Add NOTE on SubBalances missing ETH in molecule schema
- Optimize release profile: strip=true, opt-level="z", lto=true, panic="abort"
- Remove unused sha256 feature from k256 dependency

Signed-off-by: Ilja von Hoessle <4417345+iljabvh@users.noreply.github.com>

iljabvh committed:
refactor: deduplicate verification functions, add security tests

- Move verify_valid_state_sigs, verify_equal_sum_of_balances, verify_version_number to perun-common
- Remove duplicate implementations from channel-typescript and vchannel-typescript
- Remove unused alloy/sig imports from contracts after deduplication
- Add pay_eth, inflate_ckbytes, set_version mutators to test harness
- Add dispute_with_custom_sigs to Channel for signature injection tests
- Add security tests: wrong signer, corrupted signature, version regression, balance inflation, ETH payment lifecycle, channel ID determinism

Signed-off-by: Ilja von Hoessle <4417345+iljabvh@users.noreply.github.com>

iljabvh committed:
chore: remove unused imports and variables

Signed-off-by: Ilja von Hoessle <4417345+iljabvh@users.noreply.github.com>

iljabvh committed:
fix: audit findings - VC Close1 logic, owner check in merge, SUDT funding init

- Change && to || in check_valid_close1 to reject changing parents OR vcstate individually
- Add owner equality check in verify_equal_vc_status to prevent owner hijack during merge
- Pass discarded input index to verify_vc_rent_payout_merge instead of hardcoded index zero
- Fix vec![0u128, len] to vec![0u128; len] in verify_funding_in_outputs for multi-SUDT
- Update vc_merge test harness to use selected VC status as merged output

Signed-off-by: Ilja von Hoessle <4417345+iljabvh@users.noreply.github.com>

iljabvh committed:
chore: merge dev into review/size-and-tests

Resolve conflicts from the CKB Cross-Chain squash-merge by keeping the audit-branch versions: deduplicated verify_* functions in perun_common::channels and extended test mutators.

Signed-off-by: Ilja von Hoessle <4417345+iljabvh@users.noreply.github.com>

iljabvh committed:
Merge pull request #26 from perun-network/review/size-and-tests

Review/size and tests
